### ISSUE TITLE
Single-User Mode Check Combined With Database Accessibility Check Lead to Backup Failures

### ISSUE THREAD

**Issue by EdwardPollack** [olahallengren/sql-server-maintenance-solution#755]
When a database is in single-user mode, the backup script will still back it up, so long as it is accessible - even if no additional connections are available to it.  The result is that the backup fails, as does the stored procedure and the job that calls it.

In dbo.DatabaseBackup, there is a check prior to backing up a database to determine if a database is in a state where it cannot/should not be backed up.  One of those checks is this one on line 3191:
AND NOT (@CurrentUserAccess = 'SINGLE_USER' AND @CurrentIsDatabaseAccessible = 0)

This check will skip backing up a database if it is both in single-user mode and it is not accessible.  Generally speaking, few (if any) organizations want to back up databases in single-user mode and this will often fail as another process has set the database to single-user and is likely using the single connection for another task.

I would request that either:
1. These checks are separated.  This way, single-user mode databases are automatically skipped and inaccessible databases are skipped, but these 2 checks are not tied together.
2. Whether or not single-user mode databases are backed up is added as a parameter to the stored procedure.

Personally, I lean towards #1 as it makes more sense to me, but either of these options would resolve the problem.
Thank you!
-Ed

**SQL Server version and edition**
Microsoft SQL Server 2022 (RTM-GDR) (KB5021522) - 16.0.1050.5 (X64)   Jan 23 2023 17:02:42   Copyright (C) 2022 Microsoft Corporation  Developer Edition (64-bit) on Windows 10 Pro 10.0 <X64> (Build 22621: ) 
**Also Web, Standard, and Enterprise

**Version of the script**
SQL Server Maintenance Solution - SQL Server 2008, SQL Server 2008 R2, SQL Server 2012, SQL Server 2014, SQL Server 2016, SQL Server 2017, SQL Server 2019, and SQL Server 2022

**What command are you executing?**
dbo.DatabaseBackup (parameters vary per server/environment)

**What output are you getting?**
An error, whenever a database on a server is in single-user mode.

**Comment by olahallengren** [comment]
I don't understand why you getting an error here.

**Comment by olahallengren** [comment]
I don't understand why you are getting an error in this scenario. I would expect it to be skipped, if it is not accessible.

**Comment by olahallengren** [comment]
The thinking behind this logic is like this:

- If the database can be backed up, it should be backed up.
- If the database cannot be backed up because it is in single user mode and the database can't be accessed, it should be skipped. 
- If the database is not in single user mode, but can't be accessed, there is some other issue. The backup for this database will then fail.

It could be that I should introduce a parameter for whether or not single-user mode databases should be backed up.

@EdwardPollack, could you share some more information about what issue you have run into?

**Comment by Stiffboard** [comment]
I'd say skip Single user databases as it is a lottery whether the database can be backed up or not. If some other connection is current the backup will fail. Either way the job should not fail and stop, it should move on to next database. What is the logic for dbcc and index? Or is the same logic applied for those jobs?

**Comment by olahallengren** [comment]
>I'd say skip Single user databases as it is a lottery whether the database can be backed up or not. If some other connection is current the backup will fail. 

Well, in the current behavior, if there is another connection (database not accessible), then the database will be skipped.

>Either way the job should not fail and stop, it should move on to next database.

It should not fail, but I will do some testing to check what happens if it does.

**Comment by olahallengren** [comment]
I have created a pull request to skip databases in single user mode.
https://github.com/olahallengren/sql-server-maintenance-solution/pull/852

**Comment by olahallengren** [comment]
There is a related issue: #193

**Comment by EdwardPollack** [comment]
The issue with single-user databases getting backed up is that very often the single user is already in use, and the backup fails. There are 2 appropriate ways to manage this that I can come up with:

- Skip databases in single-user mode. Whatever the reason the databases are in single-user mode is out-of-scope for a backup process
- Attempt to back up single-user mode databases, but TRY/CATCH so that if it cannot establish a connection, there is not failure as a result.

It looks like the PR to skip single-user databases is already created, and that looks good!

**Comment by olahallengren** [comment]
Thank you, @EdwardPollack for the input and review. I have released it now.

**Comment by Stiffboard** [comment]
Do this selection also affect DBCC and indexoptimize? Are Single user dbs
excluded and will it cause job to fail?

Den fre 27 dec. 2024 18:18Ola Hallengren ***@***.***> skrev:

**Comment by olahallengren** [comment]
Databases in single user mode are skipped in all the stored procedures. It will not cause any jobs to fail.
